assets/js/base/context/cart-checkout/payment-methods/use-payment-method-registration.js

```diff
@@ -86,8 +89,13 @@ const usePaymentMethodRegistration = (
 				[ paymentMethod.name ]: paymentMethod,
 			};
 		};
-		for ( const paymentMethodName in registeredPaymentMethods ) {
+
+		for ( let i=0; i<PAYMENT_GATEWAY_SORT_ORDER.length; i++  ) {
+			const paymentMethodName = PAYMENT_GATEWAY_SORT_ORDER[i];
```

Comment:
There's a couple other things to consider here:

1. Express payment methods go through this logic. 

With the current loop, I think this means that some of them might not get registered. As an example, I don't think chrome pay and applePay will show up because `payment_request` is the registered name for that payment method and that identifier won't be in the `PAYMENT_GATEWAY_SORT_ORDER` array.  Currently, I'm also not aware of any ordering mechanism for Express Payment Methods in Woo Settings.

2. I think it might be good to extract the ordering out into the `usePaymentMethods` hook exported in this file.

This way the ordering only happens once based on the `PAYMENT_GATEWAY_SORT_ORDER` (you could memoize it). Then you can leave the api the way it was originally in `usePaymentMethodRegistration` and it takes care of the current problem for express payment methods.

You may need to cast the registered payment methods object (for both express payment methods and payment methods) to a Map in the process (to ensure you are preserving order). 

For example:

```js
export const usePaymentMethods = ( dispatcher ) => {
	const paymentMethods = getPaymentMethods();
	const orderedPaymentMethods = useMemo( () => {
		const ordered = new Map();
		for ( let i = 0; i < PAYMENT_GATEWAY_SORT_ORDER.length; i++ ) {
			const paymentMethodName = PAYMENT_GATEWAY_SORT_ORDER[ i ];
			const paymentMethod = paymentMethods[ paymentMethodName ];
			if ( paymentMethod ) {
				ordered.set( paymentMethodName, paymentMethods[ paymentMethodName ] );
			}
		}
	}, [ paymentMethods ] );
	return usePaymentMethodRegistration( dispatcher, orderedPaymentMethods );
}
```

Then `usePaymentMethodRegistration` would have to handle looping through the Map (which is also why you might need to convert Express payment methods to a map too). 

Or you could just convert to an array of Payment Methods as a part of the ordering methods and pass that through to `usePaymentMethodRegistration` (and just pass through `Object.values( getExpressPaymentMethods()` as the argument from `useExpressPaymentMethods`). The array would probably be simpler to work with and still would preserve order.

Reply:
Right, that seems more robust, and also good to separate out from the registration loop (which is quite a bit of code). Thanks - will get on that :)

Reply:
> With the current loop, I think this means that some of them might not get registered. As an example, I don't think chrome pay and applePay will show up

Just tested - good catch, this broke Apple Pay for me 🤕

Reply:
Started implementing this in db32784e405cafe9d05205331da0d634f7d5a5cb, using an array (rather than map) ... then realised that I need to change the interface to `usePaymentMethodRegistration` to take array, and fix up express methods too (as you noted). 

It's now infinite-looping – warning for anyone who checks out this branch!. I'm guessing because of the dispatch of available payment methods in `refreshCanMakePayments` - this ultimately changes `registeredPaymentMethods` (I'm guessing here). So probably need to move the reordering code further out, or do it "late" deep in `refreshCanMakePayments` as before.

Reply:
OK got it sorted. I've changed `usePaymentMethodRegistration` so it also takes an order param, and I'm passing this in for express and standard.

There's an extra complexity - if COD is dependent on shipping, and a non-COD-compatible shipping method is selected on page load, then COD is not present in the `PAYMENT_GATEWAY_SORT_ORDER` setting. So to handle that, I'm merging the order and all keys using lodash `union`. So you were right @nerrad - we do need to handle methods not being present in that array. 

So with that all in mind, this definitely needs a new review and thorough testing – it would be great if we can find a more elegant way of handling this. 

We may also want to fix `PAYMENT_GATEWAY_SORT_ORDER` setting server side so it always returns all methods. The fact that it's hiding COD for us is a sign that it's not the right fit for our needs.